### PR TITLE
Refactor simple_embedding sample

### PR DESCRIPTION
**Pull request by hcindyl**
Generate embedded library bytecode from the source instead of using
prebuilts. Move the bytecode module selection to the device_*.c code so
it can be target-specific.

Remove the legacy_library_loader support from `device_dylib` and
`device_local_sync`. Currently only put in the test/support for x86_64,
arm_64, and riscv_64 architectures for the test coverage. Other arch
support could be added later.

### CODE REVIEW

File: iree/samples/simple_embedding/device_dylib.c [hunk 1]
```diff
@@ -12,23 +12,24 @@
...
+  iree_hal_executable_loader_t* loader = NULL;
   iree_host_size_t loader_count = 0;
   IREE_RETURN_IF_ERROR(iree_hal_embedded_library_loader_create(
       iree_hal_executable_import_provider_null(), iree_allocator_system(),
-      &loaders[loader_count++]));
-  IREE_RETURN_IF_ERROR(iree_hal_legacy_library_loader_create(
-      iree_hal_executable_import_provider_null(), iree_allocator_system(),
-      &loaders[loader_count++]));
+      &loader));
+  loader_count++;
```
Comment (benvanik) [inline iree/samples/simple_embedding/device_dylib.c:32]: can drop loader_count (as it's always 1)

File: iree/samples/simple_embedding/device_local_sync.c [hunk 2]
```diff
@@ -12,32 +12,40 @@
...
   iree_host_size_t loader_count = 0;
   IREE_RETURN_IF_ERROR(iree_hal_embedded_library_loader_create(
       iree_hal_executable_import_provider_null(), iree_allocator_system(),
-      &loaders[loader_count++]));
-  // TODO(hcindyl): drop the use of the legacy library loader.
-  IREE_RETURN_IF_ERROR(iree_hal_legacy_library_loader_create(
-      iree_hal_executable_import_provider_null(), iree_allocator_system(),
-      &loaders[loader_count++]));
+      &loader));
+  loader_count++;
```
Comment (benvanik) [inline iree/samples/simple_embedding/device_local_sync.c:31]: drop here too

File: iree/samples/simple_embedding/simple_embedding.c [hunk 3]
```diff
@@ -16,18 +16,24 @@
...
+typedef struct iree_file_toc_t {
+  const char* name;  // the file's original name
+  const char* data;  // beginning of the file
+  size_t size;       // length of the file
+} iree_file_toc_t;
+#endif  // IREE_FILE_TOC
+
+// A function to load the vm bytecode module from the different backend targets.
+// The bytecode module is generated for the specific backend and platform.
+extern const struct iree_file_toc_t* load_bytecode_module();
```
Comment (benvanik) [inline iree/samples/simple_embedding/simple_embedding.c:35]: we should change this to return an `iree_const_byte_span_t` so we don't have the file toc leak across the boundary

File: iree/samples/simple_embedding/CMakeLists.txt [hunk 4]
```diff
@@ -53,64 +34,99 @@ iree_cc_binary(
...
-    iree::hal::vulkan::registration
+    iree::hal::local
+    iree::hal::local::loaders::embedded_library_loader
+    iree::hal::local::sync_driver
     iree::modules::hal
     iree::vm
     iree::vm::bytecode_module
 )
 
+iree_bytecode_module(
```
Comment (benvanik) [inline iree/samples/simple_embedding/CMakeLists.txt:54]: it'd be nice to have aarch64 here too so we have some kind of android coverage on the CIs

**Review by benvanik — CHANGES_REQUESTED** [review]
good cleanup! just a few minor things

File: iree/samples/simple_embedding/simple_embedding.c (hunk 3 above)
Comment (hcindyl) [reply iree/samples/simple_embedding/simple_embedding.c]: Great suggestion! Done.

File: iree/samples/simple_embedding/CMakeLists.txt (hunk 4 above)
Comment (hcindyl) [reply iree/samples/simple_embedding/CMakeLists.txt:48]: Module added. However, Android currently doesn't build sample projects (or test them). We can address it later.

File: iree/samples/simple_embedding/device_local_sync.c (hunk 2 above)
Comment (hcindyl) [reply iree/samples/simple_embedding/device_local_sync.c]: Done.

File: iree/samples/simple_embedding/device_dylib.c (hunk 1 above)
Comment (hcindyl) [reply iree/samples/simple_embedding/device_dylib.c]: Done.

**Comment by hcindyl** [comment]
Need to rebase this PR until https://github.com/google/iree/pull/6243 is merged (so lld will be available as a host binary).

**Review by benvanik — APPROVED** [review]

**Pull request merged**